tests/integration/integration_util_test.go

```diff
@@ -257,35 +259,84 @@ func runComponentSetup(t *testing.T, testDir string) {
 	synchronouslyDo(t, lockfile, timeout, fn)
 }
 
-func synchronouslyDo(t *testing.T, lockfile string, timeout time.Duration, fn func()) {
-	mutex := fsutil.NewFileMutex(lockfile)
-	defer func() {
-		assert.NoError(t, mutex.Unlock())
-	}()
+func synchronouslyDo(t testing.TB, lockfile string, timeout time.Duration, fn func()) {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
 
-	lockWait := make(chan struct{}, 1)
+	lockWait := make(chan struct{})
 	go func() {
-		for {
+		mutex := fsutil.NewFileMutex(lockfile)
+
+		// ctx.Err will be non-nil when the context finishes
+		// either because it timed out or because it got canceled.
+		for ctx.Err() == nil {
 			if err := mutex.Lock(); err != nil {
 				time.Sleep(1 * time.Second)
 				continue
 			} else {
+				defer func() {
+					assert.NoError(t, mutex.Unlock())
+				}()
 				break
 			}
 		}
 
-		fn()
-		lockWait <- struct{}{}
+		// Context may hav expired
+		// by the time we acquired the lock.
+		if ctx.Err() == nil {
+			fn()
+			close(lockWait)
+		}
 	}()
 
 	select {
-	case <-time.After(timeout):
+	case <-ctx.Done():
```

Comment:
Can we get rid of this select and _just_ wait for lockWait.
My worry is that we could start running fn() in the goroutine, then the timeout hits and this select fires, and we just leave the goroutine running and don't get to see it's result.

Reply:
Guess it means lockWait needs to be a bool of true/false for it it timed out or not.